lib/components/map/gtfs-rt-vehicle-overlay.js

```diff
@@ -0,0 +1,243 @@
+import { transit_realtime as transitRealtime } from 'gtfs-realtime-bindings'
+import L from 'leaflet'
+import memoize from 'lodash.memoize'
+import TransitVehicleOverlay from '@opentripplanner/transit-vehicle-overlay'
+import PropTypes from 'prop-types'
+import React from 'react'
+import ReactDOMServer from 'react-dom/server'
+import { FeatureGroup, MapLayer, Marker, Polyline } from 'react-leaflet'
+import styled from 'styled-components'
+
+import { getModeFromRoute } from '../../util/viewer'
+
+export const VehicleShape = styled.span`
+  background-color: #${props => props.route?.route_color || '000000'};
+  border: 1px solid #000000;
+  border-radius: 50%;
+  color: #${props => props.route?.route_text_color || 'ffffff'};
+  display: block;
+  height: 24px;
+  left: -6px;
+  line-height: 24px;
+  margin-left: auto;
+  margin-right: auto;
+  position: relative;
+  text-align: center;
+  top: -6px;
+  width: 24px;
+`
+
+/**
+ * Convert GTFS-rt entities to vehicle location from OTP-UI.
+ * @param vehicle The GTFS-rt vehicle position entity to convert.
+ * @param routes Optional GTFS-like routes list.
+ */
+function gtfsRtToTransitVehicle (gtfsRtVehicle, routes) {
+  const { position, stopId, timestamp, trip, vehicle } = gtfsRtVehicle.vehicle
+  const route = routes.find(r => r.route_id === trip.routeId)
+  let routeShortName
+  let routeType = 'BUS'
+  if (route) {
+    routeShortName = route.route_short_name
+    // Obtain the OTP route type from the GTFS route_type enum value.
+    routeType = getModeFromRoute({ type: route.route_type })
+  }
+  return {
+    heading: position.bearing,
+    id: vehicle.id,
+    lat: position.latitude,
+    lon: position.longitude,
+    reportDate: new Date(parseInt(timestamp, 10)).toLocaleString(),
+    routeShortName,
+    routeType,
+    stopId,
+    tripId: trip.tripId,
+    vehicleId: vehicle.id
+  }
+}
+
+/**
+ * A memoized function to cache the HTML icon for each route.
+ */
+const getRouteIconHtml = memoize(
+  route => ReactDOMServer.renderToStaticMarkup(
+    <VehicleShape route={route}>
+      {route.route_short_name}
+    </VehicleShape>
+  )
+)
```

Comment:
Hmmm, on second thought, I now see how the route object is used here. This won't result in any performance improvement. Each time the component mounts, the routes are fetched and then each route object gets created fresh, so when each route is passed in here that won't result in any cache hits. So, let's just remove this memoization.

Reply:
Or I can change `VehicleShape` to only take foreground and background colors as props?

Reply:
What you could do is write [a custom resolver function](https://lodash.com/docs/4.17.15#memoize) that the memoize function takes in so that something other than the route object is used as the key. Something like this:

 ```suggestion
/**
 * A memoized function to cache the HTML icon for each route.
 */
const getRouteIconHtml = memoize(
  route => ReactDOMServer.renderToStaticMarkup(
    <VehicleShape route={route}>
      {route.route_short_name}
    </VehicleShape>
  ),
  route => `${route.route_short_name}-${route.route_color}-${route.route_text_color}`
)
```

But even then, this would probably result in 0 cache hits during the first render and therefore would only be helpful if the component were mounted a second time which might not happen all that much.

Reply:
Ok, so I just removed memoization in https://github.com/opentripplanner/otp-react-redux/pull/401/commits/59763fdeb7252ca1affdd21ec2bc2572d97f0d95.